JStachio module #2864

agentgt opened the pull request:
Add jstachio support to jooby.

jknack commented:
@agentgt will review soon. In JStachio what did you do with buffer size? can you point me there?

I would like to add a event-loop buffer on jooby with a max of 8k (like you did) and reuse that buffer...

agentgt commented on modules/jooby-jstachio/src/main/java/io/jooby/jstachio/JStachioMessageEncoder.java:
```diff
@@ -0,0 +1,72 @@
+   * 
+   * @return a buffer either from a pool or a new one.
+   */
+  protected StringBuilder aquireBuffer() {
```
@jknack Per your question on buffering and potentially pooling. I use StringBuilder and no pooling at the moment. There has been lots of debate whether or not to do the whole threadlocal pooling given loom and how good newer JDKs are object creation.

agentgt commented:
I added a comment but have copied it here:

I use StringBuilder and no pooling at the moment. A new string builder initialized with a bufferSize is created everytime. 

There has been lots of debate whether or not to do the whole threadlocal pooling given loom and how good newer JDKs are object creation.

Even JTE the closest in terms of performance to JStachio removed their pooling (both JTE and mine are far faster (~ 1.7x) than Rocker):

https://github.com/casid/jte/releases/tag/2.2.0

That being said I added pretend pooling methods. Also I wasn't sure if MessageEncoder was going to change to ByteBuffer or something similar.

I guess I could try to load up whatever sandbox techempower has to test Rocker's pooling my naive StringBuilder.

Pooling is obviously going to win in isolated perf environments like JMH but from my testing using the template engine benchmark: https://github.com/agentgt/template-benchmark/ it only makes like a 10% difference on JDk 17.

jknack commented:
will look into JTE. I got and agreed on pooling+threadlocal.... the thing is: almost all the tests in tech-empower they make use and abuse of threadlocal and without them looks like jooby is not as fast, but that isn't truth.

something similar happens with router implementations... most of them are just switch/case statements :S

agentgt commented:
I can certainly add some implementation like I did for the template benchmarks: https://github.com/agentgt/template-benchmark/blob/ebaee9d71cccc41aba65c07b16ac76e1b2d29bd4/src/main/java/com/mitchellbosecke/benchmark/JStachio.java#L19

Which btw to your point on benchmarks... most forks of that template-benchmark do that little threadlocal cheat to their own library but fail to apply it to the others. That and they enable caching (compiled) but not for the others. It's one of the reasons Handlebars does not look as good on those benchmarks.

What might be better if we really need pooling is some sort of better `MessageEncoder` interface where you acquire a buffer and then release it and then I guess have that tied into the event loop buffers.

I guess the other issue w/ buffering and `MessageEncoder` is good ole team reactive. JStachio is not reactive and neither is `MessageEncoder` but team reactive requires everything be "reactive" (buffering is a form of backpressure solution but not good enough for them). I would understand the need for JStachio to be reactive if Java was lowlevel like Rust where you have intricate control of memory but most people fill up a model with plain collections (e.g. `List` instead of `Flow.Publisher`) and then render. Anyway that is my rant for the day.

agentgt commented:
@jknack let me know if you want me to squash the PR or any other changes. 

Right now the version is 0.12.0 but sometime next week it should be 1.0.0.

jknack commented:
you need to add .adoc file

agentgt commented:
Hopefully the doc is OK. It has been a long time since I have used adoc.

I'm not entirely sure how I can build a preview.

agentgt commented:
Hmm I get an artifact missing error:

```
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.1.0:java (asciidoctor) on project docs: An exception occurred while executing the Java class. asciidoctor: FAILED: /Users/agent/projects/jooby/docs/asciidoc/modules/jstachio.adoc: Failed to load AsciiDoc document: >>usage: Missing artifact: jooby-jstachio -> [Help 1]
```


I'm guessing I forgot to add jooby-jstachio to the bom or root pom (assuming adoc can do maven dependencies or that is a nifty plugin you wrote)?

agentgt commented:
Alright so now I get this error message:

```
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.1.0:java (asciidoctor) on project docs: An exception occurred while executing the Java class. File: /Users/agent/projects/jooby/docs/asciidoc/site/modules/jstachio/index.html: Cannot invoke "org.jsoup.nodes.Element.text()" because the return value of "org.jsoup.nodes.Element.selectFirst(String)" is null -> [Help 1]
```

Which I assume is some post processing you are running and missed putting something that the jsoup is expecting.

jknack commented on pom.xml:
```diff
@@ -350,6 +351,12 @@
       </dependency>
 
+      <dependency>
+        <groupId>io.jstach</groupId>
```
io.jooby

agentgt replied:
I probably should have opened intellij or eclipse instead of using vim :)

agentgt commented:
```
Caused by: java.lang.NullPointerException: Cannot invoke "org.jsoup.nodes.Element.text()" because the return value of "org.jsoup.nodes.Element.selectFirst(String)" is null
    at io.jooby.adoc.DocGenerator.languageTab (DocGenerator.java:347)
    at io.jooby.adoc.DocGenerator.document (DocGenerator.java:320)
    at io.jooby.adoc.DocGenerator.lambda$generate$6 (DocGenerator.java:122)
```

Ok I see I messed something up on the language tab.

agentgt commented:
That `{somedepVersion}` is nice. I had such a pain in the ass getting Javadoc to do similar and ended up doing what you did with a post processor right on the generated HTML itself.

agentgt commented:
@jknack Ok I think everything is working. I got the doc build to work. 

I'm not sure if JStachio currently will work with Kotlin (kapt) but I don't think that will change anything in this PR. 

Let me know if there is anything else you want me to change and/or squash.

Thanks for the help!

jknack commented:
the doc gen only expect a secondary language when `role="primary"` is present. If you removed that, then no kotlin (secondary thing) is required.

> That {somedepVersion} is nice. I had such a pain in the ass getting Javadoc to do similar and ended up doing what you did with a post processor right on the generated HTML itself.

doing doc is as hard as everything else. doc in 1.x was messy messy... This one is bit better
